### PR TITLE
Pract06:task01 | Radoslav Peev

### PR DESCRIPTION
**Pull request by Radopeev**
Radoslav Peev
SI 03 7MI0600261

### CODE REVIEW

File: task1/Deck.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#include "Spell.h"
+
+const size_t MAX_CARDS = 20;
+
+class Deck
+{
+private:
+	Monster monsterCards[MAX_CARDS];
+	Spell spellCards[MAX_CARDS];
+	size_t monsterCardsNum = 0;
```
Comment (GeorgiTerziev02) [inline task1/Deck.h:12]: monsterCardsCount is a betterName

File: task1/Deck.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	size_t spellCardsNum = 0;
+
+public:
+	Deck() = default;
+	size_t getMonsterCardsNum()const;
+	size_t getSpellCardsNum()const;
+	void addMonsterCard(const char* name,size_t attackPoints,size_t defensePoints);
+	void addSpellCard(const char* name, const char* effect, const SpellType& type);
```
Comment (GeorgiTerziev02) [inline task1/Deck.h:20]: No need of a ref on spelltype

File: task1/Deck.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	void changeMonsterCard(size_t index,const char* name, size_t attackPoints, size_t defensePoints);
+	void changeSpellCard(size_t index, const char* name, const char* effect, const SpellType& type);
```
Comment (GeorgiTerziev02) [inline task1/Deck.h:22]: same

File: task1/Deck.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	void removeMonsterCard(size_t index);
+	void removeSpellCard(size_t index);
+	Deck& operator=(const Deck& other);
```
Comment (GeorgiTerziev02) [inline task1/Deck.h:25]: why

File: task1/Deck.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+}
+void Deck::changeMonsterCard(size_t index, const char* name, size_t attackPoints, size_t defensePoints) {
+	monsterCards[index] = Monster(name,attackPoints,defensePoints);
+}
+void Deck::changeSpellCard(size_t index, const char* name, const char* effect, const SpellType& type) {
+	spellCards[index] = Spell(name, effect, type);
+}
+void Deck::removeMonsterCard(size_t index) {
+	monsterCards[monsterCardsNum] = monsterCards[index];
+	monsterCardsNum--;
```
Comment (GeorgiTerziev02) [inline task1/Deck.cpp:26]: you have to swap them

File: task1/Deck.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+}
+void Deck::removeSpellCard(size_t index) {
+	spellCards[spellCardsNum] = spellCards[index];
```
Comment (GeorgiTerziev02) [inline task1/Deck.cpp:29]: swap

File: task1/Deck.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	spellCardsNum--;
+}
+Deck& Deck::operator=(const Deck& other) {
+	for (size_t i = 0; i < monsterCardsNum; i++) {
```
Comment (GeorgiTerziev02) [inline task1/Deck.cpp:33]: y

File: task1/Spell.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	Spell() = default;
+	Spell(const char* name, const char* effect, const SpellType& type);
+
+	void setName(const char* name);
+	void setEffect(const char* effect);
+	void setType(const SpellType& type);
+
+	const char* getName() const;
+	const char* getEffect() const;
+	const SpellType& getSpellType() const;
```
Comment (GeorgiTerziev02) [inline task1/Spell.h:27]: why ref

File: task1/Duelist.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
+#include "Duelist.h"
+#include <iostream>
+#pragma warning(disable:4996)
+void Duelist::copy(const Duelist& other) {
+	strcpy(this->name, name);
+	this->deck = deck;
```
Comment (GeorgiTerziev02) [inline task1/Duelist.cpp:6]: other.name, other.deck

File: task1/Duelist.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+}
+void Duelist::removeMonsterCardInDeck(size_t index) {
+	deck.removeMonsterCard(index);
+}
+void Duelist::removeSpellCardInDeck(size_t index) {
+	deck.removeSpellCard(index);
+}
+
+Duelist::~Duelist() {
+	delete[] name;
```
Comment (GeorgiTerziev02) [inline task1/Duelist.cpp:53]: free function

File: Monster.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+public:
+	Monster() = default;
+	Monster(const char* name,size_t attackPoints,size_t defensePoints);
+	Monster(const Monster& other);
+
+	void setName(const char* name);
+	void setAttackPoints(size_t attackPoints);
+	void setDefensePoints(size_t defensePoints);
+
+	char* getName() const;
```
Comment (GeorgiTerziev02) [inline Monster.h]: const char*, do you know why

File: task1/Monster.h [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+public:
+	Monster() = default;
+	Monster(const char* name,size_t attackPoints,size_t defensePoints);
+	Monster(const Monster& other);
+
+	void setName(const char* name);
+	void setAttackPoints(size_t attackPoints);
+	void setDefensePoints(size_t defensePoints);
+
+	char* getName() const;
```
Comment (GeorgiTerziev02) [inline task1/Monster.h:21]: const char* do you know why

File: task1/Monster.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	setAttackPoints(attackPoints);
+	setDefensePoints(defensePoints);
+}
+
+Monster::Monster(const Monster& other) {
+	copy(other);
+}
+
+void Monster::setName(const char* name) {
+	strcpy(this->name, name);
```
Comment (GeorgiTerziev02) [inline task1/Monster.cpp:26]: delete the old name before setting the new

File: task1/Spell.h [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+class Spell
+{
+private:
+	char name[MAX_NAME_SYMBOLS] ="";
+	char effect[MAX_EFFECT_SYMBOLS]="";
+	SpellType type;
+public:
+	Spell() = default;
+	Spell(const char* name, const char* effect, const SpellType& type);
```
Comment (GeorgiTerziev02) [inline task1/Spell.h:19]: why ref

File: task2/Item.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+	Item(const Item& other);
+	Item& operator=(const Item& other);
+	
+	void itemCountIterator();
+	void setName(const char* name);
+	void setAvailability(bool availabilty);
+	void setPrice(float price);
+
+	size_t getItemCount()const;
+	char* getName()const;
```
Comment (GeorgiTerziev02) [inline task2/Item.h]: const

File: task2/Item.cpp [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+void Item::itemCountIterator() {
+	itemAmount++;
+}
+
+size_t Item::getItemCount()const {
+	return itemAmount;
+}
+
+void Item::setName(const char* name) {
+	this->name = new char[strlen(name) + 1];
```
Comment (GeorgiTerziev02) [inline task2/Item.cpp]: delete old

File: task2/Item.h [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+	void copyFrom(const Item& other);
+	void free();
+public:
+	Item();
+	Item(const char* name, bool availabilty, float);
+	Item(const Item& other);
+	Item& operator=(const Item& other);
+	
+	void itemCountIterator();
```
Comment (GeorgiTerziev02) [inline task2/Item.h]: bad name

File: task2/ShoppingCart.h [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+	ShoppingCart& operator=(const ShoppingCart& other);
+
+	bool addItem(const char* name, bool isAvailable, float price);
+	bool removeItem(const char* name);
+	void getItemsCount() const;
+	bool isEmpty()const;
+	void getPriceOf(const char* name)const;
+	void getTotalPrice()const;
+	void sortByName();
+	bool saveToFile(const char* fileName);
```
Comment (GeorgiTerziev02) [inline task2/ShoppingCart.h]: const

File: task2/ShoppingCart.cpp [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+void swap(Item& a, Item& b) {
+	Item temp = a;
+	a = b;
+	b = temp;
+}
+
+
+void ShoppingCart::copyFrom(const ShoppingCart& other) {
+	this->itemsCount = other.itemsCount;
+	for (int i = 0; i < other.itemsCount; i++) {
```
Comment (GeorgiTerziev02) [inline task2/ShoppingCart.cpp]: allocate the memory

File: task2/ShoppingCart.cpp [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+	items[itemsCount++].itemCountIterator();
+	std::cout << "This item has been added successfully." << std::endl;
+	return true;
+}
+bool ShoppingCart::removeItem(const char* name) {
+	int itemForRemovalIndex = find(name);
+	if (itemForRemovalIndex != -1){
+		items[itemsCount] = items[itemForRemovalIndex];
+		itemsCount--;
+		std::cout << "This item has been removed successfully." << std::endl;
```
Comment (GeorgiTerziev02) [inline task2/ShoppingCart.cpp]: Do not put these information messages inside the class

File: task2/ShoppingCart.cpp [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+bool ShoppingCart::addItem(const char* name,bool isAvailable,float price) {
+	int newItemIndex = find(name);
+	if (newItemIndex != -1) {
+		std::cout << "This item already exists." << std::endl;
+		return false;
+	}
+	if (itemsCount == capacity)
+		resize(capacity * 2);
+	items[itemsCount]=Item(name,isAvailable,price);
+	items[itemsCount++].itemCountIterator();
```
Comment (GeorgiTerziev02) [inline task2/ShoppingCart.cpp]: ?

File: task2/ShoppingCart.cpp [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+void ShoppingCart::getItemsCount() const {
+	std::cout << "There are " << itemsCount << " items at the moment in the cart." << std::endl;
+}
+bool ShoppingCart::isEmpty()const {
+	if (itemsCount == 0) {
+		std::cout << "The cart is empty." << std::endl;
+		return true;
+	}
+	std::cout << "The cart is not empty." << std::endl;
+	return false;
```
Comment (GeorgiTerziev02) [inline task2/ShoppingCart.cpp]: return itemsCount == 0

**Review by GeorgiTerziev02 — CHANGES_REQUESTED** [review]
Separate the info messages(UI) from the class

File: Monster.h (hunk 11 above)
Comment (Radopeev) [reply Monster.h]: because the value to which the pointer points is not going to be changed

**Pull request closed without merging**